### PR TITLE
Added HTTP error detail provider and refactored HTTP-sink and source package to handle error provider and fix sonar issues

### PR DESCRIPTION
**Pull request by psainics**
httpErrorManagement

### CODE REVIEW

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java [hunk 1]
```diff
@@ -37,12 +40,18 @@ public class HTTPOutputFormat extends OutputFormat<StructuredRecord, StructuredR
...
     Configuration hConf = context.getConfiguration();
     HTTPSinkConfig config = GSON.fromJson(hConf.get(CONFIG_KEY), HTTPSinkConfig.class);
-    Schema inputSchema = Schema.parseJson(hConf.get(INPUT_SCHEMA_KEY));
-    return new HTTPRecordWriter(config, inputSchema);
+    Schema inputSchema;
+    try {
+      inputSchema = Schema.parseJson(hConf.get(INPUT_SCHEMA_KEY));
+      return new HTTPRecordWriter(config, inputSchema);
+    } catch (IOException e) {
+      String errorMessage = "Unable to parse and write the record";
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java]: error Message should be unable to parse the input schema.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java [hunk 2]
```diff
@@ -37,12 +40,18 @@ public class HTTPOutputFormat extends OutputFormat<StructuredRecord, StructuredR
...
+      throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
+        errorMessage, e.getMessage(), ErrorType.UNKNOWN, true, new IOException(errorMessage));
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java]: We can wrap this IOException in IllegalState as this is because somehow schema becomes corrupt in the conf.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 3]
```diff
@@ -131,7 +134,7 @@ public void write(StructuredRecord input, StructuredRecord unused) throws IOExce
 
     if (config.getMethod().equals(REQUEST_METHOD_PUT) || config.getMethod().equals(REQUEST_METHOD_PATCH) ||
       config.getMethod().equals(REQUEST_METHOD_DELETE)
-      && !placeHolderList.isEmpty()) {
+        && !placeHolderList.isEmpty()) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: revert the changes due to indentation as it makes things confusing for the reviewer. In actions on save, reformat the file only where you are making changes

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 4]
```diff
@@ -169,76 +172,96 @@ public void checkServerTrusted(X509Certificate[] certs, String authType) {
...
-
-      if (url.getProtocol().equalsIgnoreCase("https")) {
-        // Disable SSLv3
-        System.setProperty("https.protocols", "TLSv1,TLSv1.1,TLSv1.2");
-        if (config.getDisableSSLValidation()) {
-          disableSSLValidation();
-        }
+
+      // Use try-with-resources to ensure response is closed
+      try (CloseableHttpResponse response = executeHttpRequest(httpClient, url)) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: keep a single try with resources block for httpClient and respose

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPSink.java [hunk 5]
```diff
@@ -73,13 +75,22 @@ public void prepareRun(BatchSinkContext context) {
...
     LineageRecorder lineageRecorder = new LineageRecorder(context, asset);
     lineageRecorder.createExternalDataset(context.getInputSchema());
-    List<String> fields = inputSchema == null ?
-      Collections.emptyList() :
-      inputSchema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toList());
+    List<String> fields;
+    if (inputSchema == null) {
+      fields = Collections.emptyList();
+    } else {
+      assert inputSchema.getFields() != null;
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPSink.java]: are you catching the assertion error somewhere?
if not, you can go ahead with old format, CDF has some validations for schema which will check it.

File: src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSourceConfig.java [hunk 6]
```diff
@@ -123,6 +119,10 @@ public void validateBasicAuthResponse(FailureCollector collector, HttpClient htt
           collector.addFailure(errorMessage, "Please ensure that correct credentials are provided.");
         }
       }
+    } catch (IOException e) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSourceConfig.java:125]: do we need catch block for HttpHostConnectException which was there previously

File: src/main/java/io/cdap/plugin/http/source/common/RawStringPerLine.java [hunk 7]
```diff
@@ -61,14 +64,16 @@ public boolean hasNext() {
             isLineRead = true;
             return lastLine != null;
         } catch (IOException e) { // we need to catch this, since hasNext() does not have "throws" in parent
-            throw new RuntimeException("Failed to read line from http page buffer", e);
+            String errorMessage = "Unable to read line from http page buffer";
+            throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
+              errorMessage, e.getMessage(), ErrorType.UNKNOWN, true, new IOException(errorMessage));
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/http/source/common/RawStringPerLine.java]: why are we creating a new exception when we have instance of IOException here.

File: src/test/java/io/cdap/plugin/http/etl/HttpStreamingSourceETLTest.java [hunk 8]
```diff
@@ -56,11 +56,11 @@ public class HttpStreamingSourceETLTest extends HttpSourceETLTest {
   private static final ArtifactSummary APP_ARTIFACT = new ArtifactSummary("data-streams", "1.0.0");
   private static final int WAIT_FOR_RECORDS_TIMEOUT_SECONDS = 60;
   private static final long WAIT_FOR_RECORDS_POLLING_INTERVAL_MS = 100;
+  public static final String EXPLORE_ENABLED = "explore.enabled";
 
   @ClassRule
   public static final TestConfiguration CONFIG =
-    new TestConfiguration(Constants.Explore.EXPLORE_ENABLED, false,
-                          Constants.AppFabric.SPARK_COMPAT, Compat.SPARK_COMPAT);
+    new TestConfiguration(EXPLORE_ENABLED, false, Constants.AppFabric.SPARK_COMPAT, Compat.SPARK_COMPAT);
```
Comment (vikasrathee-cs) [inline src/test/java/io/cdap/plugin/http/etl/HttpStreamingSourceETLTest.java]: why this change is required? we dont use these classes for test anymore.

File: src/main/java/io/cdap/plugin/http/source/common/RawStringPerLine.java (hunk 7 above)
Comment (Amit-CloudSufi) [reply src/main/java/io/cdap/plugin/http/source/common/RawStringPerLine.java]: updated

File: src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSourceConfig.java (hunk 6 above)
Comment (Amit-CloudSufi) [reply src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSourceConfig.java:125]: added back

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPSink.java (hunk 5 above)
Comment (Amit-CloudSufi) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPSink.java]: handled using java 8, removed assert

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java (hunk 4 above)
Comment (Amit-CloudSufi) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: updated

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java (hunk 3 above)
Comment (Amit-CloudSufi) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: reverted

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java (hunk 2 above)
Comment (Amit-CloudSufi) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java]: updated

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java (hunk 1 above)
Comment (Amit-CloudSufi) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPOutputFormat.java]: updated

**Comment by psainics** [comment]
The [Bump PR](https://github.com/data-integrations/http/pull/175) is merged, please rebase again with develop, there are conflicts 😿

**Comment by Amit-CloudSufi** [comment]
Not need anymore, raise a new PR - https://github.com/cloudsufi/http/pull/16.
closing this one
